Track milliseconds since process start in timer heap

ctiller opened the pull request:
Allows reducing a lock-then-check to an atomic load and check on the
fast path of timer checks.

Reduces locks per RPC by about 5.

ctiller commented:
Jenkins: test this please

ctiller commented:
Switching reviewers in an attempt to load balance

vjpai commented on src/core/lib/iomgr/timer_generic.c:
```diff
@@ -92,13 +113,15 @@ void grpc_timer_list_init(gpr_timespec now) {
     grpc_time_averaged_stats_init(&shard->stats, 1.0 / ADD_DEADLINE_SCALE, 0.1,
                                   0.5);
-    shard->queue_deadline_cap = now;
+    shard->queue_deadline_cap = timespec_to_atm(now);
```
Since this is already calculated as g_min_timer above, why repeat the call? (I know this isn't timing sensitive since it's initial, but still)

ctiller replied:
It's debatable, but I find this clearer: I don't need to trace more variables through the function to see that it's doing the right thing (and given this is absolutely not critical path that's a win for me).

vjpai replied:
Resolved.

vjpai commented on src/core/lib/iomgr/timer_generic.c:
```diff
@@ -178,7 +191,7 @@ void grpc_timer_init(grpc_exec_ctx *exec_ctx, grpc_timer *timer,
   GPR_ASSERT(now.clock_type == g_clock_type);
   timer->closure = closure;
-  timer->deadline = deadline;
+  timer->deadline = timespec_to_atm(deadline);
```
So timespec_to_atm seems to always round down, but the timer specification is that we won't fire before the specified deadline. In this case, we'll fire up to .999 ms before the specified deadline and that may cause some issues. I think it would be fine to round up in timespec_to_atm without violating our expectations of timers, though this change does limit the use of very high-frequency timers (which were not guaranteed to be very precise anyway)

ctiller replied:
Updated to get the rounding right. Agree on this reducing usefulness for high resolution timers: I'm prepared to live with that (and use a different technique for those timers if we need to).

Given epoll_wait() takes a millisecond timeout, I suspect we'll not notice this in practice however.

vjpai replied:
Resolved.

ctiller commented:
Updated to also include a thread local that caches g_min_time to avoid cacheline contention in very high throughput cases.

vjpai commented on src/core/lib/iomgr/timer_generic.c:
```diff
@@ -67,20 +68,58 @@ typedef struct {
+
+static gpr_atm timespec_to_atm_round_up(gpr_timespec ts) {
+  double x = GPR_MS_PER_SEC * (double)ts.tv_sec +
+             (double)ts.tv_nsec / GPR_NS_PER_MS + 1.0;
```
+0.999999999 ? To avoid adding a ms if it's an exact value

ctiller replied:
Done

ctiller commented:
Fixed a slew of bugs, added a bunch of tracing that'll hopefully help in case we ever need to dive into this code again: I'm still not convinced it's entirely right.

ctiller commented:
This is looking pretty safe to merge

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration    cpu_time    locks_per_iteration    real_time    writes_per_iteration
-------------------------------------------------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_EmptyCore                                                                                              -0.99999904677                       -1.0
BM_HpackParserParseHeader<AddIndexedSingleInternedElem>                                                                            34.5                               34.5
BM_IsolatedFilter<HttpServerFilter, SendEmptyMetadata>                                                                             9.5                                9.5
```

ctiller commented:
Now includes #10414

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                    allocs_per_iteration    atm_add_per_iteration    atm_cas_per_iteration    cpu_time     locks_per_iteration    real_time    writes_per_iteration
-----------------------------------------------------------  ----------------------  -----------------------  -----------------------  -----------  ---------------------  -----------  ----------------------
BM_EmptyCore                                                                                                  -0.999999024648                       -1.0
BM_PumpStreamClientToServer<TCP>/0                                                                            -1.94015                              -1.918635
BM_PumpStreamClientToServer<TCP>/1                                                                            -2.01425                              -2.0044
BM_PumpStreamClientToServer<TCP>/16M                                                                                                   3032223.0                           3032289.5
BM_PumpStreamClientToServer<TCP>/4k                                                                           -2.06265                              -2.07155
BM_PumpStreamClientToServer<TCP>/512                                                                          -2.0074                               -2.00648
BM_PumpStreamClientToServer<TCP>/64                                                                           -1.9815                               -1.987265
BM_PumpStreamClientToServer<TCP>/8                                                                            -1.9541                               -1.966665
BM_PumpStreamClientToServer<UDS>/0                                                                                                                  -1.01047
BM_PumpStreamClientToServer<UDS>/1                                                                                                                  -1.00772
BM_PumpStreamClientToServer<UDS>/128M                                                                         -9532.15                 -71915537.5                         -71919779.0
BM_PumpStreamClientToServer<UDS>/4k                                                                                                                 -1.10797
BM_PumpStreamClientToServer<UDS>/512                                                                                                                -1.017625
BM_PumpStreamClientToServer<UDS>/64                                                                                                                 -1.01437
BM_PumpStreamClientToServer<UDS>/8                                                                                                                  -1.009725
BM_PumpStreamServerToClient<TCP>/0                                                                            -2.22215                              -2.10801
BM_PumpStreamServerToClient<TCP>/1                                                                            -1.81295                              -1.87525
BM_PumpStreamServerToClient<TCP>/128M                                                                         -8293.3                  -97261094.0                         -97266634.0
BM_PumpStreamServerToClient<TCP>/16M                                                                                                   -5111142.0                          -5112393.0
BM_PumpStreamServerToClient<TCP>/4k                                                                           -2.09035                              -2.0711
BM_PumpStreamServerToClient<TCP>/512                                                                          -2.0229                               -2.00586
BM_PumpStreamServerToClient<TCP>/64                                                                           -2.0168                               -2.00188
BM_PumpStreamServerToClient<TCP>/8                                                                            -1.8326                               -1.888275
BM_PumpStreamServerToClient<UDS>/0                                                                                                                  -1.010695
BM_PumpStreamServerToClient<UDS>/1                                                                                                                  -1.00889
BM_PumpStreamServerToClient<UDS>/4k                                                                                                                 -1.105435
BM_PumpStreamServerToClient<UDS>/512                                                                                                                -1.01797
BM_PumpStreamServerToClient<UDS>/64                                                                                                                 -1.014565
BM_PumpStreamServerToClient<UDS>/8                                                                                                                  -1.00847
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/0/0                                                                                             -5.00015
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/0/1                                                       -9.90455                              -9.9993
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/0/2                                                       -15.0905                              -15.0
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/1/1                                                       -9.8985                               -9.99965
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/1/2                                                       -15.042                               -14.99975
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/4k/1                                                      -10.10915                             -10.00055
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/4k/2                                                      -15.1465                              -15.0003
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/512/1                                                     -10.1122                              -10.0025
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/512/2                                                     -14.971                               -14.99995
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/64/1                                                      -10.36195                             -10.00445
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/64/2                                                      -17.458                               -15.00695
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/8/1                                                       -9.8962                               -9.99845
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/8/2                                                       -15.069                               -14.99805
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/0                                                     -5.10825                              -5.0008
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/1                                                     -5.0851                  -2500.0      -5.0006                -2504.5
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/4k                                                    -5.09715                              -5.0007
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/512                                                   -5.055                                -4.99955
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/64                                                    -5.09725                              -5.003
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/8                                                     -5.04405                 -2417.5      -5.0009                -2419.5
BM_UnaryPingPong<SockPair, NoOpMutator, NoOpMutator>/0/0                                                                                            -3.9999
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/0                                                                                                 -5.00035
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/1                                                                                                 -4.99965
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/4k                                                                                                -5.0001
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/512                                                                                               -4.99755
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/64                                                                                                -5.0012
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/8                                                                                                 -5.00145
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/1/0                                                                                                 -4.9998
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/1/1                                                                                                 -5.0002
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/2M/0                                                                                   -501759.0                           -501838.5
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4k/0                                                                                                -5.0016
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4k/4k                                                                                               -4.99885
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/512/0                                                                                               -4.9605
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/512/512                                                                                             -4.9998
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/64/0                                                                                                -4.99995
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/64/64                                                                                               -4.99985
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/8/0                                                                                                 -5.0004
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/8/8                                                                                                 -4.99985
BM_UnaryPingPong<UDS, NoOpMutator, NoOpMutator>/0/0                                                                                                 -4.0
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark     allocs_per_iteration    atm_add_per_iteration      atm_cas_per_iteration  cpu_time      locks_per_iteration  real_time    writes_per_iteration
------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_EmptyCore                                                                     -1.00                              -1.00
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark     allocs_per_iteration    atm_add_per_iteration      atm_cas_per_iteration  cpu_time      locks_per_iteration  real_time    writes_per_iteration
------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_EmptyCore                                                                     -1.00                              -1.00
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark     allocs_per_iteration    atm_add_per_iteration      atm_cas_per_iteration  cpu_time      locks_per_iteration  real_time    writes_per_iteration
------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_EmptyCore                                                                     -1.00                              -1.00
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark     allocs_per_iteration    atm_add_per_iteration      atm_cas_per_iteration  cpu_time      locks_per_iteration  real_time    writes_per_iteration
------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_EmptyCore                                                                     -1.00                              -1.00
```

grpc-jenkins commented:
```
Performance differences noted:
Benchmark     allocs_per_iteration    atm_add_per_iteration      atm_cas_per_iteration  cpu_time      locks_per_iteration  real_time    writes_per_iteration
------------  ----------------------  -----------------------  -----------------------  ----------  ---------------------  -----------  ----------------------
BM_EmptyCore                                                                     -1.00                              -1.00
```
